consensus/raft/raft.go

```diff
@@ -427,9 +435,134 @@ func (rw *raftWrapper) Peers() ([]string, error) {
 	return ids, nil
 }
 
-// only call when Raft is shutdown
-func (rw *raftWrapper) Clean() error {
-	dbh := newDataBackupHelper(rw.dataFolder)
+// latestSnapshot looks for the most recent raft snapshot stored at the
+// provided basedir.  It returns a boolean indicating if any snapshot is
+// readable, the snapshot's metadata, and a reader to the snapshot's bytes
+func latestSnapshot(raftDataFolder string) (bool, *hraft.SnapshotMeta, io.ReadCloser, error) {
+	store, err := hraft.NewFileSnapshotStore(raftDataFolder, RaftMaxSnapshots, nil)
+	if err != nil {
+		return false, nil, nil, err
+	}
+	snapMetas, err := store.List()
+	if err != nil {
+		return false, nil, nil, err
+	}
+	if len(snapMetas) == 0 { // no error if snapshot isn't found
+		return false, nil, nil, nil
+	}
+	meta, r, err := store.Open(snapMetas[0].ID)
+	if err != nil {
+		return false, nil, nil, err
+	}
+	return true, meta, r, nil
+}
+
+// LastState decodes the latest snapshot into the provided state
+func LastState(cfg *Config, state state.State) (bool, error) {
+	// Read most recent snapshot
+	dataFolder, err := makeDataFolder(cfg.BaseDir, cfg.DataFolder)
+	if err != nil {
+		return false, err
+	}
+	found, _, r, err := latestSnapshot(dataFolder)
+	if err != nil {
+		return false, err
+	}
+	if !found {
+		return false, nil
+	}
+
+	// !!!!!!Use Decodesnapshot instead of unmarshal
+	rawBytes, err := ioutil.ReadAll(r)
+	if err != nil {
+		return true, err
+	}
+	err = p2praft.DecodeSnapshot(rawBytes, state)
+	if err != nil {
+		return true, err
+	}
+	return true, nil
+}
+
+// SnapshotMigrate saves a raft snapshot containing a migrated version of the
+// latest snapshot's state.  If no state yet exists there is nothing to migrate
+// and it returns an error.  Old snapshots are cleaned up.
+// Only call when Raft is shutdown.
+func SnapshotMigrate(cfg *Config, newState state.State) error {
```

Comment:
Do you think it would make sense to do the migrate in main.go. I mean, instead of calling SnapshotMigrate() and let this (which is part of consensus) do everything, do instead:

* `consensus.LastState(cfg, st) []byte` to load the last snapshot
* `st.Restore([]byte)` to migrate it
* `consensus.SaveState(cfg,st)` to write a new snapshot

Or is there something strongly preventing this?

Reply:
Thanks for this.  For some reason I thought there was something strongly preventing this but there actually isn't anything.  I am pretty moving to this will address all of your other comments.

Reply:
This opens up a bit more the question of how unmarshal handles an old version of snapshot.  The current plan is to record the version, throw away the data and return no errors.  If you'd rather throw an error I am ok with that too.